src/sqlfluff/utils/reflow/elements.py

```diff
@@ -0,0 +1,383 @@
+
+
+@dataclass(frozen=True)
+class ReflowElement:
+    """Base reflow element class."""
+
+    segments: Sequence[RawSegment]
+
+    @staticmethod
+    def _class_types(segments: Sequence[RawSegment]) -> Set[str]:
+        return set(chain.from_iterable(seg.class_types for seg in segments))
+
+    @property
+    def class_types(self):
+        """The set of contained class types.
+
+        Parallel to BaseSegment.class_types
+        """
+        return self._class_types(self.segments)
+
+
+@dataclass(frozen=True)
+class ReflowBlock(ReflowElement):
+    """Class for keeping track of elements to reflow.
+
+    It holds segments to reflow and also exposes configuration
+    around how they are expected to reflow around others.
+
+    The attributes exposed are designed to be "post configuration"
+    i.e. they should reflect configuration appropriately.
+
+    NOTE: These are the smallest unit of "work" within
+    the reflow methods, and may contain meta segments.
+    """
+
+    # Options for spacing rules are:
+    # - single:         the default (one single space)
+    # - touch:          no whitespace
+    # - any:            don't enforce any spacing rules
+    spacing_before: str
+    spacing_after: str
+    # The depth info is used in determining where to put line breaks.
+    depth_info: DepthInfo
+    # This stores relevant configs for segments in the stack.
+    stack_spacing_configs: Dict[int, str]
+
+    @classmethod
+    def from_config(cls, segments, config: ReflowConfig, depth_info: DepthInfo):
+        """Extendable constructor which accepts config."""
+        block_config = config.get_block_config(cls._class_types(segments))
+        # Populate any spacing_within config.
+        # TODO: This needs decent unit tests - not just what happens in rules.
+        stack_spacing_configs = {}
+        for hash, class_types in zip(
+            depth_info.stack_hashes, depth_info.stack_class_types
+        ):
+            spacing_within = config.get_block_config(class_types).get(
+                "spacing_within", None
+            )
+            if spacing_within:
+                stack_spacing_configs[hash] = spacing_within
+        return cls(
+            segments=segments,
+            spacing_before=block_config.get("spacing_before", "single"),
+            spacing_after=block_config.get("spacing_after", "single"),
+            depth_info=depth_info,
+            stack_spacing_configs=stack_spacing_configs,
+        )
+
+
+@dataclass(frozen=True)
+class ReflowPoint(ReflowElement):
+    """Class for keeping track of editable elements in reflow.
+
+    It holds segments which can be changed during a reflow operation
+    such as whitespace and newlines.
+
+    It holds no configuration and is influenced by the blocks either
+    side.
+    """
+
+    def respace_point(
+        self,
+        prev_block: Optional[ReflowBlock] = None,
+        next_block: Optional[ReflowBlock] = None,
+        fixes: Optional[List[LintFix]] = None,
```

Comment:
Nitpick: Is it useful to make `fixes` optional, since the list could be empty?

Reply:
Yeah I reckon we can do that.